# Pré-processamento — Dataset CSEDM / ProgSnap2 v6

Transforma os dados brutos do CSEDM em sequências KT prontas para BKT, DKT e Code-DKT.  
Metodologia: fase de *Data Preparation* do EDM Process (Kalita et al., 2025).

---

## Dataset e Splits

O CSEDM disponibiliza dois conjuntos de splits com critérios de elegibilidade distintos. É essencial entender qual usar para reproduzir os resultados do paper.

### Split primário — Spring 2019 / "All" (este notebook)

**Protocolo Shi et al. (2022)** — único split compatível com reprodução das métricas do paper:

| Parâmetro | Valor |
|-----------|-------|
| Arquivo | `data/CSEDM/MainTable.csv` |
| Critério de inclusão | `min_attempts >= 3` (Run.Program globais, todos os assignments) |
| Estudantes após filtro | **410** (de 413 brutos) |
| Split | `train_test_split(students, test_size=0.2, random_state=1)` |
| Treino | **328 estudantes** |
| Teste | **82 estudantes** |
| Assignments disponíveis no test set | **5** — A439, A487, A492, A494, A502 |
| Taxa de corretos (Run.Program, Score==1.0) | **23.68%** — match com paper |

**Quando usar:** para treinar e avaliar BKT, DKT e Code-DKT. Toda comparação com Table 1 e Table 2 de Shi et al. (2022) deve usar este split.

### Split Release/ — CSEDM Data Challenge 2021 (não usar para modelagem KT)

O diretório `Release/` (removido do pipeline) era o split do *CSEDM Data Challenge 2021*, uma competição de predição acadêmica. Apresenta diferenças críticas que o tornam incompatível com a reprodução do paper:

| Parâmetro | Valor |
|-----------|-------|
| Critério de inclusão | Alunos que "completaram o curso" |
| Estudantes | **329** (subconjunto de 413, critério diferente) |
| Split | 75/25 (≠ 80/20 do paper) |
| Assignments no Release/Test | **apenas A1–A3** (A4–A5 eram *prediction targets* da competição) |

**Quando NÃO usar:** para modelagem KT comparável com Shi et al. (2022). A ausência de A4–A5 no test set impede a replicação de 2 dos 5 pontos das tabelas do paper. O critério de inclusão diferente também altera a distribuição de proficiência dos estudantes.

**Referência:** Shi et al. (2022) — Section 4.1; Price et al. (2020) — ProgSnap2 v6; Pankiewicz, Shi & Baker (2025).

In [1]:
import sys
from pathlib import Path

import numpy as np
import pandas as pd

SEED = 42
np.random.seed(SEED)

NOTEBOOK_DIR = Path().resolve()
ROOT = NOTEBOOK_DIR.parent
DATA_ROOT = ROOT / 'data' / 'CSEDM'
RESULTS_DIR = ROOT / 'results'
RESULTS_DIR.mkdir(exist_ok=True)

# Adicionar src/ ao path para importar data_loader
SRC = ROOT / 'src'
if str(SRC) not in sys.path:
    sys.path.insert(0, str(SRC))

from data_loader import load_main_table, load_spring2019_split

print(f'SEED = {SEED}')
print(f'DATA_ROOT = {DATA_ROOT}')
print(f'DATA_ROOT existe: {DATA_ROOT.exists()}')

SEED = 42
DATA_ROOT = /home/leokuntz/Documents/repositories/studies/tcc.edm.kt/data/CSEDM
DATA_ROOT existe: True


---

## 1 — Setup e Carregamento dos Dados

### 1.1 — Carregamento e split dos dados

**Contexto:** O pré-processamento começa carregando `data/CSEDM/MainTable.csv` e aplicando o split 80/20 do protocolo Shi et al. (2022). A função `load_spring2019_split` encapsula o filtro `min_attempts >= 3` (contado em Run.Program globais), o `train_test_split(students, test_size=0.2, random_state=1)` e a normalização de tipos (ServerTimestamp → UTC, AssignmentID/ProblemID → Int64).  
**Hipótese:** O train set deve ter 328 estudantes, o test set 82, e todos os 5 assignments devem estar presentes no test set (A439, A487, A492, A494, A502). A taxa de corretos (Run.Program com Score==1.0) deve ser ~23.68% — match com o paper.  
**Referência:** Shi et al. (2022) — Section 4.1, "23.68% of the attempts are correct"; split 80/20 com random_state=1.

In [2]:
# Carregar e aplicar split 80/20 (Spring 2019, protocolo Shi et al. 2022)
train_df, test_df = load_spring2019_split(DATA_ROOT, test_size=0.2, random_state=1, min_attempts=3)
# load_labels() removido: Release/ eliminado; labels em data/early.csv e data/late.csv se necessário

splits = {
    'train': train_df,
    'test':  test_df,
}

print(f'{"Split":<8} {"Eventos":>10} {"Estudantes":>12} {"Assignments":>13}')
print('-' * 48)
for name, df in splits.items():
    n_events   = len(df)
    n_subjects = df['SubjectID'].nunique()
    n_assign   = df['AssignmentID'].nunique()
    print(f'{name:<8} {n_events:>10,} {n_subjects:>12,} {n_assign:>13,}')

Split       Eventos   Estudantes   Assignments
------------------------------------------------
train       164,184          328             5
test         37,375           82             5


In [3]:
# Validação do split — asserts de reprodutibilidade (Shi et al. 2022 protocol)
train_run = train_df[train_df['EventType'] == 'Run.Program']
print(f"Train: {train_df['SubjectID'].nunique()} alunos | corretos: {(train_run['Score']==1.0).mean()*100:.2f}%")
print(f"Test:  {test_df['SubjectID'].nunique()} alunos")
print(f"Assignments test: {sorted(test_df['AssignmentID'].dropna().unique().tolist())}")

assert train_df['SubjectID'].nunique() == 328, f"Esperado 328 alunos train; encontrado {train_df['SubjectID'].nunique()}"
assert test_df['SubjectID'].nunique() == 82, f"Esperado 82 alunos test; encontrado {test_df['SubjectID'].nunique()}"
assert set(test_df['AssignmentID'].dropna().unique()) == {439, 487, 492, 494, 502}, \
    f"Esperado todos os 5 assignments no test set; encontrado: {sorted(test_df['AssignmentID'].dropna().unique().tolist())}"

train_rate = (train_run['Score'] == 1.0).mean()
assert abs(train_rate - 0.2368) < 0.005, f"Taxa inesperada: {train_rate:.4f} (esperado ~0.2368 ±0.005)"

print()
print(f"OK — split reproduzível: 328 treino + 82 teste, 5 assignments, taxa {train_rate:.4%}")

Train: 328 alunos | corretos: 23.79%
Test:  82 alunos
Assignments test: [439, 487, 492, 494, 502]

OK — split reproduzível: 328 treino + 82 teste, 5 assignments, taxa 23.7882%


**Achado:** Split 80/20 reproduzível — train com 328 estudantes, test com 82, todos os 5 assignments presentes no test set. Taxa de corretos no train ~23.68%, match com o paper Shi et al. (2022).  
**Implicação para modelagem:** Toda comparação com os resultados do paper usa este split (random_state=1). Comparação direta com Table 1 e Table 2 de Shi et al. (2022) agora válida.

### 1.2 — Benchmark de reprodutibilidade

**Contexto:** Shi et al. (2022) reporta que o dataset de treinamento contém 23.68% de tentativas corretas (Run.Program com Score == 1.0). Confirmar este valor é o primeiro teste de reprodutibilidade do pipeline.  
**Hipótese:** A taxa calculada no train split deve ser ~23.68%, refletindo o mesmo filtro `min_attempts >= 3` usado no paper.  
**Referência:** Shi et al. (2022) — Section 4.1, "23.68% of the attempts are correct".

In [4]:
runs_train = train_df[train_df['EventType'] == 'Run.Program'].copy()
runs_test  = test_df[test_df['EventType']  == 'Run.Program'].copy()

correct_rate_train = (runs_train['Score'] == 1.0).mean()
correct_rate_test  = (runs_test['Score']  == 1.0).mean()

print(f'Train split — Run.Program: {len(runs_train):,} eventos')
print(f'  Taxa de corretos (Score == 1.0): {correct_rate_train:.4%}')
print(f'  Paper Shi et al. (2022):          23.6800%')
print(f'  Divergência:                      {abs(correct_rate_train - 0.2368):.4%}')
print()
print(f'Test split — Run.Program: {len(runs_test):,} eventos')
print(f'  Taxa de corretos (Score == 1.0): {correct_rate_test:.4%}')

Train split — Run.Program: 56,423 eventos
  Taxa de corretos (Score == 1.0): 23.7882%
  Paper Shi et al. (2022):          23.6800%
  Divergência:                      0.1082%

Test split — Run.Program: 13,200 eventos
  Taxa de corretos (Score == 1.0): 23.2424%


**Achado:** A taxa de corretos no train split é ~23.68%, match com o valor reportado por Shi et al. (2022) — benchmark de reprodutibilidade confirmado.  
**Implicação para modelagem:** O dataset está íntegro para reprodução do paper. A baixa proporção de acertos (~23.7%) justifica o uso de AUC como métrica primária em vez de acurácia — uma predição trivial "sempre errado" teria acurácia de 76.3%, mas AUC de ~50%.

### 1.3 — Estrutura e tipos de dados

**Contexto:** Verificar se as colunas críticas estão presentes e com tipos corretos. Os modelos KT dependem de `SubjectID`, `AssignmentID`, `ProblemID`, `ServerTimestamp` (ordem cronológica), `EventType` e `Score`.  
**Hipótese:** Todas as colunas críticas devem estar presentes no train split com os tipos corretos após `load_spring2019_split`.  
**Referência:** Price et al. (2020) — ProgSnap2 v6 specification.

In [5]:
CRITICAL_COLS = ['SubjectID', 'AssignmentID', 'ProblemID', 'ServerTimestamp',
                 'EventType', 'Score', 'CodeStateID']

print('=== Train split — dtypes e nulos ===')
info = pd.DataFrame({
    'dtype':  train_df.dtypes.astype(str),
    'nulls':  train_df.isna().sum(),
    'null_%': (train_df.isna().mean() * 100).round(2),
})
print(info.to_string())
print()

missing = [c for c in CRITICAL_COLS if c not in train_df.columns]
if missing:
    print(f'AVISO — colunas críticas ausentes: {missing}')
else:
    print('OK — todas as colunas críticas presentes no train split')

# Score só existe em Run.Program — nulo em Compile e Compile.Error é esperado
score_null_outside_run = train_df[
    (train_df['EventType'] != 'Run.Program') & train_df['Score'].notna()
]
print(f'Score não-nulo fora de Run.Program: {len(score_null_outside_run)} (esperado: 0)')

=== Train split — dtypes e nulos ===
                                         dtype   nulls  null_%
Order                                    int64       0    0.00
SubjectID                                  str       0    0.00
ToolInstances                              str       0    0.00
ServerTimestamp            datetime64[us, UTC]       0    0.00
ServerTimezone                             str       0    0.00
CourseID                                   str       0    0.00
CourseSectionID                          int64       0    0.00
AssignmentID                             Int64       0    0.00
ProblemID                                Int64       0    0.00
CodeStateID                                str       0    0.00
IsEventOrderingConsistent                 bool       0    0.00
EventType                                  str       0    0.00
Score                                  float64  107761   65.63
Compile.Result                             str  107761   65.63
CompileMessageType

**Achado:** Todas as colunas críticas estão presentes. `Score` é nulo em `Compile` e `Compile.Error` (comportamento esperado — Score só existe em `Run.Program`). `ServerTimestamp` foi convertido para `datetime64[ns, UTC]` pelo `load_main_table`, garantindo ordenação cronológica correta.  
**Implicação para modelagem:** Nenhuma engenharia de colunas adicional é necessária para o carregamento. O `load_main_table` de `src/data_loader.py` entrega o DataFrame pronto para as etapas de filtragem (tarefa 2).

---

## 2 — Filtragem por Modelo

### 2.1 — filter_for_bkt_dkt e filter_for_code_dkt

**Contexto:** BKT e DKT processam apenas sequências de acerto/erro sem informação do código-fonte; portanto, entram na sequência somente eventos `Run.Program` (execuções com Score). Code-DKT, por outro lado, extrai embeddings AST a cada submissão — incluindo código com erros de compilação — e por isso adiciona os eventos `Compile.Error` como `correct=0`. Essa distinção de protocolo é central para reproduzir os resultados de Shi et al. (2022) e Pankiewicz, Shi & Baker (2025).  
**Hipótese:** Após a filtragem, BKT/DKT devem conter apenas `Run.Program` (~46k eventos em Release/Train); Code-DKT deve conter `Run.Program` + `Compile.Error` (~87k eventos). A proporção de corretos em Code-DKT deve ser menor que em BKT/DKT, pois os `Compile.Error` contribuem exclusivamente com `correct=0`.  
**Referência:** Shi et al. (2022) — Section 3.1, "correct=1 when all tests pass"; Pankiewicz, Shi & Baker (2025) — srcML-DKT, EDM 2025.

In [6]:
from data_loader import filter_for_bkt_dkt, filter_for_code_dkt

# --- Filtragem ---
bkt_dkt_train = filter_for_bkt_dkt(train_df)
bkt_dkt_test  = filter_for_bkt_dkt(test_df)

code_dkt_train = filter_for_code_dkt(train_df)
code_dkt_test  = filter_for_code_dkt(test_df)

# --- Verificação de EventType: nenhum outro tipo deve ter passado ---
assert set(bkt_dkt_train['EventType'].unique()) == {'Run.Program'}, "ERRO: EventType inesperado em bkt_dkt_train"
assert set(bkt_dkt_test['EventType'].unique())  == {'Run.Program'}, "ERRO: EventType inesperado em bkt_dkt_test"
assert set(code_dkt_train['EventType'].unique()).issubset({'Run.Program', 'Compile.Error'}), "ERRO: EventType inesperado em code_dkt_train"
assert set(code_dkt_test['EventType'].unique()).issubset({'Run.Program', 'Compile.Error'}),  "ERRO: EventType inesperado em code_dkt_test"
print("OK — assertions de EventType passaram em todos os splits")

# --- Estatísticas dos filtros ---
print()
print(f"{'Split':<25} {'Eventos':>10} {'EventTypes':>30} {'Taxa correto':>14}")
print('-' * 85)

for name, df in [
    ('BKT/DKT — train',   bkt_dkt_train),
    ('BKT/DKT — test',    bkt_dkt_test),
    ('Code-DKT — train',  code_dkt_train),
    ('Code-DKT — test',   code_dkt_test),
]:
    event_types = ', '.join(sorted(df['EventType'].unique()))
    correct_rate = df['correct'].mean()
    print(f"{name:<25} {len(df):>10,} {event_types:>30} {correct_rate:>14.4%}")

# --- Detalhe de EventType em Code-DKT ---
print()
print("Code-DKT — train: contagem por EventType")
print(code_dkt_train['EventType'].value_counts().to_string())
print()
print("Code-DKT — test: contagem por EventType")
print(code_dkt_test['EventType'].value_counts().to_string())

# --- Confirmar que colunas 'correct' são binárias ---
assert bkt_dkt_train['correct'].isin([0, 1]).all(),  "correct não é binário em bkt_dkt_train"
assert code_dkt_train['correct'].isin([0, 1]).all(), "correct não é binário em code_dkt_train"
print()
print("OK — coluna 'correct' é binária (0/1) em ambos os filtros")

OK — assertions de EventType passaram em todos os splits

Split                        Eventos                     EventTypes   Taxa correto
-------------------------------------------------------------------------------------
BKT/DKT — train               56,423                    Run.Program       23.7882%
BKT/DKT — test                13,200                    Run.Program       23.2424%
Code-DKT — train             107,761     Compile.Error, Run.Program       12.4553%
Code-DKT — test               24,175     Compile.Error, Run.Program       12.6908%

Code-DKT — train: contagem por EventType


EventType
Run.Program      56423
Compile.Error    51338

Code-DKT — test: contagem por EventType
EventType
Run.Program      13200
Compile.Error    10975

OK — coluna 'correct' é binária (0/1) em ambos os filtros


**Achado:** `filter_for_bkt_dkt` retém apenas `Run.Program` no train e test splits. `filter_for_code_dkt` adiciona os `Compile.Error` (correct=0). Todas as assertions de EventType passam; `correct` é binária (0/1) em ambos os casos.  
**Implicação para modelagem:** As sequências BKT/DKT partem de ~23.68% de corretos; Code-DKT parte de valor menor porque inclui os Compile.Error como incorretos. A métrica AUC (e não acurácia) é essencial em ambos os cenários. A decisão de incluir Compile.Error no Code-DKT é empírica e teórica: esses eventos carregam informação da evolução do código que o srcML consegue parsear (Pankiewicz et al., 2025).

---

## 3 — Construção de Sequências KT

### 3.1 — build_sequences()

**Contexto:** Os modelos BKT, DKT e Code-DKT recebem como entrada uma sequência de tentativas de um estudante em um assignment, ordenada cronologicamente. A função `build_sequences` transforma o DataFrame filtrado em listas de sequências per-student, adicionando a flag `is_first_attempt` necessária para calcular o *first-attempt AUC* — a métrica primária de Shi et al. (2022). Processar por assignment é essencial porque os modelos são treinados independentemente para cada assignment (protocolo KC = ProblemID por assignment).  
**Hipótese:** A estrutura de sequências deve refletir as características vistas na EDA: Assignment A1 tem o maior número de estudantes com sequências completas; comprimentos médios variam entre assignments. A flag `is_first_attempt` deve cobrir todos os pares (SubjectID, ProblemID) únicos — um por estudante por problema.  
**Referência:** Shi et al. (2022) — Section 3 "Sequence truncation at 50"; Pankiewicz, Shi & Baker (2025).

In [7]:
from data_loader import build_sequences

# AssignmentIDs reais no dataset (não são 1–5 sequenciais)
assignment_ids = sorted(bkt_dkt_train['AssignmentID'].dropna().unique().tolist())
print(f'AssignmentIDs disponíveis no train split: {assignment_ids}')

# Construir sequências para BKT/DKT — primeiro assignment como exemplo de verificação
a1_id = assignment_ids[0]
seqs_bkt_a1 = build_sequences(bkt_dkt_train, assignment_id=a1_id)

print(f'\nBKT/DKT — train split — AssignmentID={a1_id}')
print(f'  Número de sequências (estudantes): {len(seqs_bkt_a1)}')

seq_lengths = [len(s['events']) for s in seqs_bkt_a1]
print(f'  Comprimento mínimo: {min(seq_lengths)}')
print(f'  Comprimento médio:  {sum(seq_lengths)/len(seq_lengths):.1f}')
print(f'  Comprimento máximo: {max(seq_lengths)}')

# Verificar is_first_attempt: deve ter exatamente 1 True por (SubjectID, ProblemID)
for seq in seqs_bkt_a1:
    df_s = seq['events']
    first_counts = df_s.groupby('ProblemID')['is_first_attempt'].sum()
    assert (first_counts == 1).all(), (
        f"Estudante {seq['subject_id']}: is_first_attempt inválido para problemas {first_counts[first_counts != 1].index.tolist()}"
    )
print('  OK — is_first_attempt: exatamente 1 True por (SubjectID, ProblemID)')

# Verificar ordenação cronológica dentro de cada sequência
for seq in seqs_bkt_a1:
    ts = seq['events']['ServerTimestamp']
    assert ts.is_monotonic_increasing, f"Sequência de {seq['subject_id']} não está ordenada cronologicamente"
print('  OK — ordenação cronológica verificada em todas as sequências')

# --- Exemplo de sequência (primeiros 6 eventos do primeiro estudante) ---
print()
print(f'=== Exemplo: primeiros 6 eventos da sequência do estudante 0 (A={a1_id}) ===')
example = seqs_bkt_a1[0]
print(f"subject_id:    {example['subject_id']}")
print(f"assignment_id: {example['assignment_id']}")
display_cols = ['ProblemID', 'ServerTimestamp', 'EventType', 'correct', 'is_first_attempt']
print(example['events'][display_cols].head(6).to_string(index=False))

# --- Contagem de sequências por assignment (BKT/DKT e Code-DKT) ---
print()
print(f'{"AssignmentID":<14} {"BKT/DKT seqs":>14} {"Code-DKT seqs":>15}')
print('-' * 46)
for aid in assignment_ids:
    n_bkt = len(build_sequences(bkt_dkt_train, aid))
    n_code = len(build_sequences(code_dkt_train, aid))
    print(f'{aid:<14} {n_bkt:>14,} {n_code:>15,}')

# --- Verificar que is_first_attempt cobre todos os pares únicos (SubjectID, ProblemID) ---
all_first = sum(
    seq['events']['is_first_attempt'].sum()
    for seq in seqs_bkt_a1
)
expected_pairs = bkt_dkt_train[bkt_dkt_train['AssignmentID'] == a1_id] \
    .groupby(['SubjectID', 'ProblemID']).ngroups
assert all_first == expected_pairs, (
    f"is_first_attempt count ({all_first}) != unique (SubjectID, ProblemID) pairs ({expected_pairs})"
)
print(f'\nOK — is_first_attempt cobre exatamente {expected_pairs} pares únicos (SubjectID, ProblemID) em A={a1_id}')

AssignmentIDs disponíveis no train split: [439, 487, 492, 494, 502]

BKT/DKT — train split — AssignmentID=439
  Número de sequências (estudantes): 307
  Comprimento mínimo: 4
  Comprimento médio:  39.0
  Comprimento máximo: 209
  OK — is_first_attempt: exatamente 1 True por (SubjectID, ProblemID)
  OK — ordenação cronológica verificada em todas as sequências

=== Exemplo: primeiros 6 eventos da sequência do estudante 0 (A=439) ===
subject_id:    00c54f9462673d4c09d2a88121860841
assignment_id: 439
 ProblemID           ServerTimestamp   EventType  correct  is_first_attempt
        13 2019-02-24 04:27:49+00:00 Run.Program        0              True
        13 2019-02-24 04:29:44+00:00 Run.Program        0             False
        13 2019-02-24 04:33:35+00:00 Run.Program        0             False
        13 2019-02-24 04:33:49+00:00 Run.Program        0             False
        13 2019-02-24 04:34:50+00:00 Run.Program        1             False
       232 2019-02-24 04:40:53+00:00 Run.P

439                       307             307
487                       272             272
492                       290             290


494                       253             253
502                       245             245

OK — is_first_attempt cobre exatamente 2949 pares únicos (SubjectID, ProblemID) em A=439


**Achado:** `build_sequences` constrói as sequências corretamente — a flag `is_first_attempt` cobre exatamente um evento por (SubjectID, ProblemID), a ordenação cronológica é garantida em todas as sequências, e o número de sequências varia por assignment (A1 com mais estudantes que A4/A5, refletindo o padrão de desistência visto na EDA). Estrutura de cada dicionário de sequência:

| Campo | Tipo | Descrição |
|-------|------|-----------|
| `subject_id` | `str` | Identificador anônimo do estudante |
| `assignment_id` | `int` | ID do assignment (1–5) |
| `events` | `pd.DataFrame` | Tentativas ordenadas por `ServerTimestamp` |

Colunas do DataFrame `events`:

| Coluna | Tipo | Descrição |
|--------|------|-----------|
| `SubjectID` | `str` | Identificador do estudante (redundante com `subject_id`) |
| `AssignmentID` | `Int64` | ID do assignment |
| `ProblemID` | `Int64` | ID do problema — KC do modelo |
| `ServerTimestamp` | `datetime64[UTC]` | Timestamp UTC da submissão |
| `EventType` | `str` | `'Run.Program'` (BKT/DKT) ou `{'Run.Program', 'Compile.Error'}` (Code-DKT) |
| `correct` | `int` | Label binária: 1 se `Run.Program` com `Score==1.0`, 0 caso contrário |
| `CodeStateID` | `str` | Referência ao estado do código (para extrair AST no Code-DKT) |
| `is_first_attempt` | `bool` | **True** para a primeira tentativa do estudante no problema (cronologicamente) |

**Implicação para modelagem:** A flag `is_first_attempt` permite calcular *first-attempt AUC* filtrando `events[events['is_first_attempt']]` após obter as predições do modelo — exatamente o protocolo de Shi et al. (2022). A estrutura em lista de dicionários (uma entrada por estudante) é diretamente iterável pelo loop de treinamento dos modelos DKT e Code-DKT, e convertível para os formatos esperados por pyBKT.

---

## 4 — Truncagem e Validação

### 4.1 — truncate_sequences()

**Contexto:** Shi et al. (2022) truncam cada sequência de estudante nas **últimas 50 tentativas** para limitar o custo computacional do LSTM e reduzir o viés de estudantes com históricos muito longos (outliers de engajamento). A truncagem mantém os eventos mais recentes porque eles refletem o estado de conhecimento mais próximo do momento de avaliação — e porque a maioria dos estudantes tem sequências menores que 50 (EDA: mediana ~37 tentativas em A1 BKT/DKT), tornando a truncagem efetiva apenas para os casos extremos.  
**Hipótese:** A maior parte das sequências (>50%) não será truncada; apenas os estudantes mais engajados (sequências longas) serão afetados. Após a truncagem, a taxa de corretos **deve aumentar** em relação à sequência completa, pois os eventos mais recentes correspondem à fase mais avançada do aprendizado — onde os estudantes erram menos. Para Code-DKT, o efeito deve ser ainda mais pronunciado porque os Compile.Error são mais frequentes no início da sequência.  
**Referência:** Shi et al. (2022) — Section 3, "We truncate student sequences to the last 50 attempts".

In [8]:
from data_loader import truncate_sequences

MAX_LEN = 50

# --- Construir sequências completas para todos os assignments ---
seqs_bkt_all  = {}  # assignment_id -> list[dict]
seqs_code_all = {}

for aid in assignment_ids:
    seqs_bkt_all[aid]  = build_sequences(bkt_dkt_train,  assignment_id=aid)
    seqs_code_all[aid] = build_sequences(code_dkt_train, assignment_id=aid)

# --- Truncar ---
seqs_bkt_trunc  = {aid: truncate_sequences(seqs, MAX_LEN) for aid, seqs in seqs_bkt_all.items()}
seqs_code_trunc = {aid: truncate_sequences(seqs, MAX_LEN) for aid, seqs in seqs_code_all.items()}

# --- Assertion principal: nenhuma sequência pode ter mais de MAX_LEN eventos ---
for aid, seqs in seqs_bkt_trunc.items():
    assert all(len(seq['events']) <= MAX_LEN for seq in seqs), \
        f"BKT/DKT A={aid}: sequência com mais de {MAX_LEN} eventos após truncagem"
for aid, seqs in seqs_code_trunc.items():
    assert all(len(seq['events']) <= MAX_LEN for seq in seqs), \
        f"Code-DKT A={aid}: sequência com mais de {MAX_LEN} eventos após truncagem"
print(f"OK — assert all(len(seq['events']) <= {MAX_LEN} ...) passa em todos os assignments e modelos")

# --- Estatísticas de truncagem por assignment ---
print()
print("=== BKT/DKT ===")
print(f"{'AssignmentID':<14} {'Seqs':>6} {'Truncadas':>10} {'% trunc':>9} "
      f"{'Len média (antes)':>18} {'Len média (depois)':>19}")
print('-' * 82)

for aid in assignment_ids:
    full  = seqs_bkt_all[aid]
    trunc = seqs_bkt_trunc[aid]
    n_trunc = sum(1 for s in full if len(s['events']) > MAX_LEN)
    avg_before = sum(len(s['events']) for s in full)  / len(full)
    avg_after  = sum(len(s['events']) for s in trunc) / len(trunc)
    print(f"{aid:<14} {len(full):>6} {n_trunc:>10} {n_trunc/len(full):>9.1%} "
          f"{avg_before:>18.1f} {avg_after:>19.1f}")

print()
print("=== Code-DKT ===")
print(f"{'AssignmentID':<14} {'Seqs':>6} {'Truncadas':>10} {'% trunc':>9} "
      f"{'Len média (antes)':>18} {'Len média (depois)':>19}")
print('-' * 82)

for aid in assignment_ids:
    full  = seqs_code_all[aid]
    trunc = seqs_code_trunc[aid]
    n_trunc = sum(1 for s in full if len(s['events']) > MAX_LEN)
    avg_before = sum(len(s['events']) for s in full)  / len(full)
    avg_after  = sum(len(s['events']) for s in trunc) / len(trunc)
    print(f"{aid:<14} {len(full):>6} {n_trunc:>10} {n_trunc/len(full):>9.1%} "
          f"{avg_before:>18.1f} {avg_after:>19.1f}")

# --- Taxa de corretos: antes vs depois da truncagem ---
print()
print("=== Taxa de corretos: antes vs depois da truncagem (Release/Train, todos os assignments) ===")

def correct_rate_from_seqs(seqs_by_aid):
    all_correct = sum(
        seq['events']['correct'].sum()
        for seqs in seqs_by_aid.values() for seq in seqs
    )
    all_total = sum(
        len(seq['events'])
        for seqs in seqs_by_aid.values() for seq in seqs
    )
    return all_correct / all_total if all_total > 0 else float('nan')

rate_bkt_before  = correct_rate_from_seqs(seqs_bkt_all)
rate_bkt_after   = correct_rate_from_seqs(seqs_bkt_trunc)
rate_code_before = correct_rate_from_seqs(seqs_code_all)
rate_code_after  = correct_rate_from_seqs(seqs_code_trunc)

print(f"{'Modelo':<14} {'Antes trunc':>14} {'Após trunc':>12} {'Δ':>8}")
print('-' * 52)
print(f"{'BKT/DKT':<14} {rate_bkt_before:>14.4%} {rate_bkt_after:>12.4%} {rate_bkt_after - rate_bkt_before:>+8.4%}")
print(f"{'Code-DKT':<14} {rate_code_before:>14.4%} {rate_code_after:>12.4%} {rate_code_after - rate_code_before:>+8.4%}")

# --- Verificar is_first_attempt após truncagem ---
for aid, seqs in seqs_bkt_trunc.items():
    for seq in seqs:
        events = seq['events']
        # Deve haver exatamente 1 is_first_attempt por ProblemID na janela truncada
        first_counts = events.groupby('ProblemID')['is_first_attempt'].sum()
        assert (first_counts == 1).all(), (
            f"BKT/DKT A={aid} subject={seq['subject_id']}: "
            f"is_first_attempt inválido após truncagem"
        )
print()
print("OK — is_first_attempt recalculado corretamente após truncagem em BKT/DKT")


OK — assert all(len(seq['events']) <= 50 ...) passa em todos os assignments e modelos

=== BKT/DKT ===
AssignmentID     Seqs  Truncadas   % trunc  Len média (antes)  Len média (depois)
----------------------------------------------------------------------------------
439               307         76     24.8%               39.0                31.8
487               272         93     34.2%               47.2                34.9
492               290         96     33.1%               48.0                31.7
494               253         60     23.7%               39.6                32.5
502               245         42     17.1%               31.3                26.4

=== Code-DKT ===
AssignmentID     Seqs  Truncadas   % trunc  Len média (antes)  Len média (depois)
----------------------------------------------------------------------------------
439               307        175     57.0%               93.2                40.6
487               272        176     64.7%               


OK — is_first_attempt recalculado corretamente após truncagem em BKT/DKT


**Achado:** A assertion `all(len(seq['events']) <= 50 ...)` passa sem erro em todos os assignments e modelos. A proporção de sequências efetivamente truncadas é baixa em BKT/DKT (a maioria dos estudantes tem <50 Run.Program); em Code-DKT a truncagem afeta mais sequências porque Compile.Error events são adicionados, aumentando o comprimento médio. A taxa de corretos **aumenta após a truncagem** em BKT/DKT (os eventos mais recentes refletem o aprendizado acumulado — estudantes erram menos no final). Para Code-DKT, o efeito é menor porque Compile.Error events (sempre `correct=0`) aparecem ao longo de toda a sequência, e a janela final ainda pode conter erros de compilação.

**Divergência em relação ao valor bruto:** A taxa bruta de Release/Train em Run.Program é 23.70% (benchmark Shi et al., 2022). Após truncagem, a taxa sobe levemente porque eventos do início da sequência — onde a taxa de acertos é menor (fase de familiarização com o problema) — são removidos. Essa divergência é esperada e desejável: o modelo é treinado na janela mais recente (estado atual de conhecimento), não no histórico completo.  
**Implicação para modelagem:** A truncagem não distorce a métrica *first-attempt AUC* (calculada sobre `is_first_attempt`, que é recalculado corretamente na janela truncada). O efeito principal é reduzir o custo de memória do LSTM (sequências de comprimento fixo ≤50) e eliminar o viés de estudantes com históricos muito longos que dominariam o gradiente.

---

## 5 — Serialização dos Artefatos

### 5.1 — Salvar sequências processadas em disco

**Contexto:** Após filtragem, construção e truncagem, as sequências KT precisam ser serializadas em disco para que os notebooks de modelagem (04–06) possam carregá-las diretamente sem re-executar o pipeline de pré-processamento. A serialização também serve como ponto de reprodutibilidade: qualquer colaborador pode re-executar `02_preprocessing.ipynb` e obter artefatos idênticos (SEED=42 fixado).
**Hipótese:** Dois artefatos distintos são necessários — `sequences_bkt_dkt.pkl` e `sequences_code_dkt.pkl` — porque os protocolos de filtragem diferem: BKT/DKT usam apenas `Run.Program`; Code-DKT inclui `Compile.Error`. Cada artefato armazena os splits `train` e `test` já truncados (máx. 50 eventos/sequência), indexados por `AssignmentID`.
**Referência:** Shi et al. (2022) — protocolo de split e truncagem; Pankiewicz, Shi & Baker (2025) — inclusão de Compile.Error no Code-DKT.

In [9]:
import pickle

# --- Construir e truncar sequências de teste ---
seqs_bkt_test_all  = {aid: build_sequences(bkt_dkt_test,  assignment_id=aid) for aid in assignment_ids}
seqs_code_test_all = {aid: build_sequences(code_dkt_test, assignment_id=aid) for aid in assignment_ids}

seqs_bkt_test_trunc  = {aid: truncate_sequences(seqs, MAX_LEN) for aid, seqs in seqs_bkt_test_all.items()}
seqs_code_test_trunc = {aid: truncate_sequences(seqs, MAX_LEN) for aid, seqs in seqs_code_test_all.items()}

# Verificar que todos os 5 assignments têm sequências no test set (Spring 2019 completo)
test_aids_bkt  = [aid for aid, seqs in seqs_bkt_test_trunc.items()  if seqs]
test_aids_code = [aid for aid, seqs in seqs_code_test_trunc.items() if seqs]
print(f'AssignmentIDs com sequências no test split — BKT/DKT: {test_aids_bkt}')
print(f'AssignmentIDs com sequências no test split — Code-DKT: {test_aids_code}')
assert set(test_aids_bkt) == {439, 487, 492, 494, 502}, \
    f"Esperado todos os 5 assignments no test; encontrado: {test_aids_bkt}"

# --- Montar artefatos ---
artifact_bkt_dkt = {
    'train':          seqs_bkt_trunc,
    'test':           seqs_bkt_test_trunc,
    'assignment_ids': assignment_ids,
    'max_len':        MAX_LEN,
    'seed':           SEED,
    'description':    'Sequências KT para BKT e DKT — apenas Run.Program, correct=(Score==1.0)',
}

artifact_code_dkt = {
    'train':          seqs_code_trunc,
    'test':           seqs_code_test_trunc,
    'assignment_ids': assignment_ids,
    'max_len':        MAX_LEN,
    'seed':           SEED,
    'description':    'Sequências KT para Code-DKT — Run.Program + Compile.Error (correct=0)',
}

# --- Serializar ---
path_bkt  = RESULTS_DIR / 'sequences_bkt_dkt.pkl'
path_code = RESULTS_DIR / 'sequences_code_dkt.pkl'

with open(path_bkt, 'wb') as f:
    pickle.dump(artifact_bkt_dkt, f, protocol=pickle.HIGHEST_PROTOCOL)
with open(path_code, 'wb') as f:
    pickle.dump(artifact_code_dkt, f, protocol=pickle.HIGHEST_PROTOCOL)

print(f'\nArtefatos salvos em {RESULTS_DIR}/')
print(f'  sequences_bkt_dkt.pkl  — {path_bkt.stat().st_size / 1024:.1f} KB')
print(f'  sequences_code_dkt.pkl — {path_code.stat().st_size / 1024:.1f} KB')

# --- Estatísticas detalhadas ---
def artifact_stats(artifact, label):
    print(f'\n=== {label} ===')
    for split_name in ('train', 'test'):
        seqs_by_aid = artifact[split_name]
        all_lengths = [len(seq['events']) for seqs in seqs_by_aid.values() for seq in seqs]
        if not all_lengths:
            print(f'  {split_name}: sem sequências')
            continue
        total_seqs   = sum(len(seqs) for seqs in seqs_by_aid.values())
        total_events = sum(all_lengths)
        all_correct  = sum(
            seq['events']['correct'].sum()
            for seqs in seqs_by_aid.values() for seq in seqs
        )
        aids_with_data = [aid for aid, seqs in seqs_by_aid.items() if seqs]
        unique_students = len({seq['subject_id'] for seqs in seqs_by_aid.values() for seq in seqs})
        print(f'  {split_name}:')
        print(f'    Assignments c/ dados:  {len(aids_with_data)} ({aids_with_data})')
        print(f'    Estudantes únicos:     {unique_students:,}')
        print(f'    Sequências total:      {total_seqs:,}')
        print(f'    Comprimento mín:       {min(all_lengths)}')
        print(f'    Comprimento médio:     {sum(all_lengths)/len(all_lengths):.1f}')
        print(f'    Comprimento máx:       {max(all_lengths)}')
        print(f'    Eventos total:         {total_events:,}')
        print(f'    Taxa de corretos:      {all_correct/total_events:.4%}')
        print(f'    Por assignment:')
        print(f'    {"AID":<8} {"Seqs":>6} {"Len min":>8} {"Len méd":>9} {"Len máx":>8}')
        print(f'    {"-"*46}')
        for aid, seqs in seqs_by_aid.items():
            if not seqs:
                print(f'    {aid:<8} {0:>6} {"—":>8} {"—":>9} {"—":>8}')
                continue
            lens = [len(s['events']) for s in seqs]
            print(f'    {aid:<8} {len(seqs):>6} {min(lens):>8} {sum(lens)/len(lens):>9.1f} {max(lens):>8}')

artifact_stats(artifact_bkt_dkt,  'BKT/DKT')
artifact_stats(artifact_code_dkt, 'Code-DKT')

# --- Schema keys ---
print('\n=== Schema dos artefatos (chaves de nível superior) ===')
for k, v in artifact_bkt_dkt.items():
    if isinstance(v, dict):
        print(f'  {k!r:<20} dict[int -> list[dict]]  # {len(v)} assignments')
    elif isinstance(v, list):
        print(f'  {k!r:<20} list[int]  # {v}')
    else:
        print(f'  {k!r:<20} {type(v).__name__}  # {v!r}')

print('\n=== Schema de cada sequência (dict interno) ===')
example_seq = artifact_bkt_dkt['train'][assignment_ids[0]][0]
for k, v in example_seq.items():
    if hasattr(v, 'dtypes'):
        print(f'  {k!r:<20} pd.DataFrame  # {len(v)} linhas, colunas: {list(v.columns)}')
    else:
        print(f'  {k!r:<20} {type(v).__name__}  # {v!r}')

AssignmentIDs com sequências no test split — BKT/DKT: [439, 487, 492, 494, 502]
AssignmentIDs com sequências no test split — Code-DKT: [439, 487, 492, 494, 502]



Artefatos salvos em /home/leokuntz/Documents/repositories/studies/tcc.edm.kt/results/
  sequences_bkt_dkt.pkl  — 11718.0 KB
  sequences_code_dkt.pkl — 13101.3 KB

=== BKT/DKT ===
  train:
    Assignments c/ dados:  5 ([439, 487, 492, 494, 502])
    Estudantes únicos:     328
    Sequências total:      1,367
    Comprimento mín:       1
    Comprimento médio:     31.5
    Comprimento máx:       50
    Eventos total:         43,110
    Taxa de corretos:      28.6778%
    Por assignment:
    AID        Seqs  Len min   Len méd  Len máx
    ----------------------------------------------
    439         307        4      31.8       50
    487         272        2      34.9       50
    492         290        5      31.7       50
    494         253        1      32.5       50
    502         245        1      26.4       50
  test:
    Assignments c/ dados:  5 ([439, 487, 492, 494, 502])
    Estudantes únicos:     82
    Sequências total:      338
    Comprimento mín:       3
    Comprimento

**Achado:** Os dois artefatos foram serializados com sucesso. Todos os 5 assignments têm sequências no test set — confirmação de que o Spring 2019 completo (80/20 split) cobre A494 e A502 no teste, diferentemente do Release/ que tinha apenas A1–A3.

**Estatísticas consolidadas — BKT/DKT (`sequences_bkt_dkt.pkl`):**

| Split | Estudantes | Sequências | Comp. mín | Comp. médio | Comp. máx |
|-------|-----------|-----------|-----------|------------|----------|
| train | 328 | 1.367 | 1 | 31,5 | 50 |
| test  | 82  | 338   | 3 | 32,2 | 50 |

**Estatísticas consolidadas — Code-DKT (`sequences_code_dkt.pkl`):**

| Split | Estudantes | Sequências | Comp. mín | Comp. médio | Comp. máx |
|-------|-----------|-----------|-----------|------------|----------|
| train | 328 | 1.367 | 1 | 39,0 | 50 |
| test  | 82  | 338   | 3 | 39,9 | 50 |

O Code-DKT apresenta comprimento médio maior (39,0 vs 31,5 no treino) porque inclui eventos `Compile.Error` na sequência, aumentando o número de eventos por estudante antes da truncagem em 50.

**Schema dos artefatos `sequences_bkt_dkt.pkl` e `sequences_code_dkt.pkl`:**

| Chave | Tipo | Descrição |
|-------|------|----------|
| `'train'` | `dict[int → list[dict]]` | Sequências truncadas de treino, indexadas por `AssignmentID` |
| `'test'` | `dict[int → list[dict]]` | Sequências truncadas de teste, indexadas por `AssignmentID` |
| `'assignment_ids'` | `list[int]` | IDs dos 5 assignments: `[439, 487, 492, 494, 502]` |
| `'max_len'` | `int` | Comprimento máximo das sequências: `50` |
| `'seed'` | `int` | Seed de reprodutibilidade: `42` |
| `'description'` | `str` | Texto descritivo do artefato |

Cada sequência (item da lista interna) é um dicionário com:

| Chave | Tipo | Descrição |
|-------|------|----------|
| `'subject_id'` | `str` | ID anônimo do estudante |
| `'assignment_id'` | `int` | ID do assignment |
| `'events'` | `pd.DataFrame` | Eventos ordenados cronologicamente, máx. 50 linhas |

Colunas de `events`: `SubjectID`, `AssignmentID`, `ProblemID`, `ServerTimestamp`, `EventType`, `correct` (int 0/1), `CodeStateID`, `is_first_attempt` (bool), além de colunas auxiliares.

**Implicação para modelagem:** Os notebooks de modelagem (04–06) carregam os artefatos com `pickle.load`. A flag `is_first_attempt` está preservada no `events` DataFrame, permitindo calcular *first-attempt AUC* sem re-executar o pré-processamento. Com todos os 5 assignments no test set, a comparação com Table 1 e Table 2 de Shi et al. (2022) agora é direta.